src/sdk/pynni/nni/nas/pytorch/cdarts/trainer.py

```diff
@@ -0,0 +1,203 @@
+import json
+import logging
+import os
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+import apex
+from apex.parallel import DistributedDataParallel
+from nni.nas.pytorch.cdarts import RegularizedDartsMutator, RegularizedMutatorParallel, DartsDiscreteMutator
+from nni.nas.pytorch.utils import AverageMeterGroup
+from utils import CyclicIterator, TorchTensorEncoder, accuracy, reduce_metrics
+
+PHASE_SMALL = "small"
+PHASE_LARGE = "large"
+
+class InteractiveKLLoss(nn.Module):
+    def __init__(self, temperature):
+        super().__init__()
+        self.temperature = temperature
+        # self.kl_loss = nn.KLDivLoss(reduction = 'batchmean')
+        self.kl_loss = nn.KLDivLoss()
+
+    def forward(self, student, teacher):
+        return self.kl_loss(F.log_softmax(student / self.temperature, dim=1),
+                            F.softmax(teacher / self.temperature, dim=1))
+
+
+class CdartsTrainer(object):
+    def __init__(self, model_small, model_large, criterion, loaders, samplers, logger, config):
+        train_loader, valid_loader = loaders
+        train_sampler, valid_sampler = samplers
+        self.train_loader = CyclicIterator(train_loader, train_sampler, config.distributed)
```

Comment:
CyclicIterator is in example utils but referenced here. I think the utils necessary to trainer should be moved here. To test, you need to install from source code. Follow instructions [here](https://github.com/microsoft/nni/blob/master/docs/en_US/Tutorial/SetupNniDeveloperEnvironment.md).

Reply:
revision done.